tscat_gui/__init__.py

```diff
@@ -51,8 +51,180 @@ def __init__(self, parent=None):
+        self.state.state_changed.connect(self.__state_changed)
 
+    def __state_changed(self, action: str, type: Union[Type[tscat._Catalogue], Type[tscat._Event]],
+                        uuids: Sequence[str]) -> None:
+        if action in ['changed', 'moved', 'inserted', 'deleted', 'active_select', 'passive_select']:
+            if type == tscat._Catalogue:
+                if action not in ['active_select', 'passive_select']:
+                    self.catalogue_model.reset()
+
+                indexes = list(map(self.catalogue_model.index_from_uuid, uuids))
+                indexes = list(map(self.catalogue_sort_filter_model.mapFromSource, indexes))
+                self.programmatic_select = True
+                self.catalogues_view.clearSelection()
+                for index in indexes:
+                    self.catalogues_view.selectionModel().select(index,
+                                                                 QtCore.QItemSelectionModel.SelectionFlag.Select)
+                self.programmatic_select = False
+            else:
+                if action not in ['active_select', 'passive_select']:
+                    self.events_model.reset()
+
+                indexes = list(map(self.events_model.index_from_uuid, uuids))
+                indexes = list(map(self.events_sort_model.mapFromSource, indexes))
+                self.programmatic_select = True
+                self.events_view.clearSelection()
+                for index in indexes:
+                    self.events_view.selectionModel().select(index,
+                                                             QtCore.QItemSelectionModel.SelectionFlag.Select |
+                                                             QtCore.QItemSelectionModel.SelectionFlag.Rows)
+                self.programmatic_select = False
+
+        if action == 'active_select':
+            self.move_to_trash_action.setEnabled(False)
+            self.restore_from_trash_action.setEnabled(False)
+            self.delete_action.setEnabled(False)
+            self.new_event_action.setEnabled(False)
+            self.export_action.setEnabled(False)
+
+            if uuids:
+                if len(uuids) == 1:
+                    self.new_event_action.setEnabled(True)
+
+                enable_restore = False
+                enable_move_to_trash = False
+                for entity in map(get_entity_from_uuid_safe, uuids):
+                    if entity.is_removed():
+                        enable_restore |= True
+                    else:
+                        enable_move_to_trash |= True
+
+                self.restore_from_trash_action.setEnabled(enable_restore)
+                self.move_to_trash_action.setEnabled(enable_move_to_trash)
+                self.delete_action.setEnabled(True)
+                self.export_action.setEnabled(True)
+
+    def __current_event_changed(self, _: QtCore.QModelIndex, __: QtCore.QModelIndex):
+        if self.programmatic_select:
```

Comment:
an if not seems way better

Reply:
I love early returns. Much better than a whole tree of `if <true-conditions>` (not the case here I admit) and indented code.

What's the problem?

Reply:
It both doesn't give any performance benefit nor improves readability. The important piece of code is not the "return None" but the one actually doing something here (L120 and 121). So I don't see why you want to make it more complicated by making implicit the fact that it should only be evaluated when self.programmatic_select is False.

Reply:
So it's a question of taste and habits. I'll change it to the style you prefer.